### PR TITLE
CT 116: Allow for new preview version to be generated without uploading a file #19667

### PR DESCRIPTION
**Pull request by zurbergram**
## Description
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/19667

### CODE REVIEW

**Review by shawkey-daniel — APPROVED** [review]
LGTM

**Review by sonntag-adam — APPROVED** [review]

File: app/models/upload.rb [hunk 1]
```diff
@@ -72,6 +72,18 @@ def self.last_uploads_rows
     uploads.sort_by { |upload| upload.csv_type.downcase }
   end
 
+  def self.latest_uploads
+    preview = Version.current_preview
+    production = Version.current_production
+    return if production.blank?
+
+    created_at = preview.present? && preview.number == production.number ? production.created_at : preview.created_at
```
Comment (jholton) [inline app/models/upload.rb]: What is supposed to happen if `preview.present?` returns `false`?

File: app/models/upload.rb [hunk 2]
```diff
@@ -72,6 +72,18 @@ def self.last_uploads_rows
     uploads.sort_by { |upload| upload.csv_type.downcase }
   end
 
+  def self.latest_uploads
```
Comment (jholton) [inline app/models/upload.rb]: Can we get some specs for this method please?

File: app/views/dashboards/index.html.erb [hunk 3]
```diff
@@ -23,18 +23,43 @@
...
-          data: { confirm: 'Are you sure you want to create a new preview version?' },
-          class: "btn btn-warning btn-xs",
-          role: "button",
-          id: "version_build"
+    <div id="preview_dialog" title="Generate New Preview Version">
+      <div id="preview_dialog_div">
+        <p class="generate_preview_question">
+          <b>Are you sure you want to create a new preview version?</b>
+        </p>
+        <% if @latest_uploads.length > 0 %>
```
Comment (jholton) [inline app/views/dashboards/index.html.erb]: `@latest_uploads` will be `nil` if `Version.current_production` is `nil` (see `Upload.latest_uploads`)

**Review by jholton — CHANGES_REQUESTED** [review]

File: app/models/upload.rb (hunk 1 above)
Comment (zurbergram) [reply app/models/upload.rb]: Fixed logic

File: app/models/upload.rb (hunk 2 above)
Comment (zurbergram) [reply app/models/upload.rb]: Created

File: app/views/dashboards/index.html.erb (hunk 3 above)
Comment (zurbergram) [reply app/views/dashboards/index.html.erb]: Method returns empty array now if production is blank

File: app/models/upload.rb (hunk 2 above)
Comment (jholton) [inline app/models/upload.rb]: Can we rename this?  I just noticed there is a `.last_uploads` too (used in this method), and so I naturally wonder what the difference is between `last uploads` and `latest uploads`.
How about we rename this something like.  `since_last_version_creation` (`Upload.since_last_version_creation`)?  Thoughts?

File: app/models/upload.rb [hunk 4]
```diff
@@ -72,6 +72,18 @@ def self.last_uploads_rows
...
+    preview = Version.current_preview
+    production = Version.current_production
+    return [] if production.blank?
+
+    created_at = preview.present? && preview.number > production.number ? preview.created_at : production.created_at
+
+    Upload.last_uploads
```
Comment (jholton) [inline app/models/upload.rb]: You don't have to convert this into an array first.  You can do it all with SQL:
`Upload.last_uploads.where("updated_at > ?", created_at)`

File: app/models/upload.rb (hunk 2 above)
Comment (zurbergram) [reply app/models/upload.rb]: renamed

File: app/models/upload.rb (hunk 4 above)
Comment (zurbergram) [reply app/models/upload.rb]: changed

**Review by jholton — APPROVED** [review]
LGTM

**Pull request merged**